src/libfetchers/git.cc

```diff
@@ -24,6 +27,212 @@ static std::string readHead(const Path & path)
     return chomp(runProgram("git", true, { "-C", path, "rev-parse", "--abbrev-ref", "HEAD" }));
 }
 
+static string getRootTree(const Path & path, const string & gitrev) {
+    auto commitOutput = runProgram("git", true, { "-C", path, "cat-file", "-p", gitrev });
+    auto lines = tokenizeString<std::vector<string>>(commitOutput, "\n");
+    auto words = tokenizeString<std::vector<string>>(lines[0]);
+    auto rootTree = words[1];
+    return rootTree;
+}
+
+enum gitType {
+  blob,
+  tree,
+  commit
+};
+
+string getBlob(const Path & path, const string & gitrev, const string & blobhash) {
+  return runProgram("git", true, { "-C", path, "cat-file", "-p", blobhash });
+}
+
+std::pair<bool, std::string> getActualUrl(const Input &input) {
+  // file:// URIs are normally not cloned (but otherwise treated the
+  // same as remote URIs, i.e. we don't use the working tree or
+  // HEAD). Exception: If _NIX_FORCE_HTTP is set, or the repo is a bare git
+  // repo, treat as a remote URI to force a clone.
+  static bool forceHttp = getEnv("_NIX_FORCE_HTTP") == "1"; // for testing
+  auto url = parseURL(getStrAttr(input.attrs, "url"));
+  bool isBareRepository =
+      url.scheme == "file" && !pathExists(url.path + "/.git");
+  bool isLocal = url.scheme == "file" && !forceHttp && !isBareRepository;
+  return {isLocal, isLocal ? url.path : url.base};
+}
+
+std::string getGitDir(const std::string & actualUrl) {
+  auto gitDir = actualUrl + "/.git";
+  auto commonGitDir = chomp(runProgram(
+      "git", true, {"-C", actualUrl, "rev-parse", "--path-format=absolute", "--git-common-dir"}));
+  if (commonGitDir != ".git")
+    gitDir = commonGitDir;
+
+  return gitDir;
+}
+
+/* Check whether this repo has any commits. There are
+   probably better ways to do this. */
```

Comment:
Strictly speaking there's a hilarious but more correct answer:
```
git cat-file --batch-check='%(objecttype)' --batch-all-objects --unordered | grep commit | head -1
```
But I'm sure any sane repo is in fact going to have an entry in `/refs/heads`.